hub/scheduler.py

```diff
@@ -0,0 +1,204 @@
+
+
+def get_due_tasks() -> list[ScheduledRun]:
+    """Get tasks that are due to run."""
+    try:
+        now = datetime.now()
+
+        with get_session() as session:
+            due_tasks = session.exec(
+                select(ScheduledRun).where(ScheduledRun.run_at <= now, ScheduledRun.has_run == False)  # noqa: E712
+            ).all()
+
+            return list(due_tasks) if due_tasks else []
+    except Exception as e:
+        logger.error(f"Error getting due tasks: {e}")
+        return []
+
+
+async def main():
+    """Main function to set up and run the scheduler."""
+    logger.info("Starting scheduler process")
+
+    # Set up signal handlers for graceful shutdown
+    loop = asyncio.get_running_loop()
+
+    for sig in (signal.SIGINT, signal.SIGTERM):
+        loop.add_signal_handler(sig, lambda s=sig: asyncio.create_task(shutdown(s)))
+
+    # Load auth token
+    auth_token = load_auth_token()
+
+    # Get configuration from environment variables
+    read_scheduled_runs = os.getenv("READ_SCHEDULED_RUNS", "false").lower() == "true"
+    read_near_events = os.getenv("READ_NEAR_EVENTS", "false").lower() == "true"
+    read_x_events = os.getenv("READ_X_EVENTS", "false").lower() == "true"
+
+    logger.info(
+        f"Events configuration: Scheduled runs: {read_scheduled_runs}, NEAR: {read_near_events}, X: {read_x_events}"
+    )
+
+    scheduler = get_async_scheduler()
+
+    # Set up scheduled jobs
+    if read_scheduled_runs:
+        job = partial(process_due_tasks, auth_token)
+        scheduler.add_job(job, IntervalTrigger(seconds=1), name="schedule_run")
+        logger.info("Added scheduled runs job")
+
+    if read_near_events:
+        process_near_events_initial_state()
+        job = partial(near_events_task, auth_token)
+        scheduler.add_job(job, IntervalTrigger(seconds=1), name="near_events")
+        logger.info("Added near events job")
+
+    if read_x_events:
+        job = partial(x_events_task, auth_token)
+        # immediately run the job
+        await job()
+        scheduler.add_job(job, IntervalTrigger(seconds=120), name="x_events")
+        logger.info("Added X events job")
+
+    if read_near_events or read_x_events or read_scheduled_runs:
+        scheduler.start()
+        logger.info("Scheduler started")
+
+        # Keep the event loop running
+        while True:
+            await asyncio.sleep(1)
+    else:
+        logger.warning("No jobs configured. Scheduler will exit.")
+
+
+async def shutdown(signal):
+    """Handle graceful shutdown."""
+    logger.info(f"Received exit signal {signal.name}...")
+    scheduler = get_async_scheduler()
+
+    if scheduler.running:
+        logger.info("Shutting down scheduler...")
+        scheduler.shutdown()
+
+    tasks = [t for t in asyncio.all_tasks() if t is not asyncio.current_task()]
+
+    if tasks:
+        logger.info(f"Cancelling {len(tasks)} outstanding tasks")
+        for task in tasks:
+            task.cancel()
```

Comment:
Should we wait for `task.done()` so we don't fail any tasks in progress?

Reply:
> Using task.done() checks would create race conditions - they only report status but don't wait for completion. gather() actively awaits termination, ensuring all tasks finish cleanup (even if they ignore CancelledError). Your current code is safer.

Reply:
We would still keep the `gather()` and wait for completion.  `task.done(0` wasn't right – but ideally we could give in-progress tasks a few seconds to complete gracefully before hitting the `cancel()`.